### ISSUE TITLE
unable to use a GraphQLEnumType as a mutation parameter

### ISSUE THREAD

**Issue by trunkate** [Enigmatis/graphql-java-annotations#130]
I tried to use a Java Enum as a parameter of mutation method without success.

I had to recreate an EnumFunction TypeFunction like the one you removed in the latest release to make it working.

Everything seems fine, but maybe I'm doing something completely wrong! Any suggestion?

```
public class EnumFunction implements TypeFunction {
    private final Map<String, GraphQLTypeReference> processing = new ConcurrentHashMap<>();
    private final Map<String, GraphQLType> types = new ConcurrentHashMap<>();

    @Override
    public String getTypeName(Class<?> aClass, AnnotatedType annotatedType) {
        GraphQLName name = aClass.getAnnotation(GraphQLName.class);
        return graphql.annotations.processor.util.NamingKit.toGraphqlName(name == null ? aClass.getSimpleName() : name.value());
    }

    @Override
    public boolean canBuildType(Class<?> aClass, AnnotatedType annotatedType) {
        return Enum.class.isAssignableFrom(aClass);
    }

    @Override
    public GraphQLType buildType(boolean input, Class<?> aClass, AnnotatedType annotatedType, ProcessingElementsContainer container) {
        return buildType(aClass, annotatedType, container);
    }

    @Override
    public GraphQLType buildType(Class<?> aClass, AnnotatedType annotatedType, ProcessingElementsContainer container) {
        String typeName = getTypeName(aClass, annotatedType);
        if (types.containsKey(typeName)) {
            return types.get(typeName);
        } else if (processing.containsKey(typeName)) {
            return processing.getOrDefault(typeName, new GraphQLTypeReference(typeName));
        } else {

            processing.put(typeName, new GraphQLTypeReference(typeName));

            //noinspection unchecked
            Class<? extends Enum> enumClass = (Class<? extends Enum>) aClass;
            GraphQLEnumType.Builder builder = GraphQLEnumType.newEnum();
            builder.name(typeName);

            GraphQLDescription description = aClass.getAnnotation(GraphQLDescription.class);
            if (description != null) {
                builder.description(description.value());
            }

            List<Enum> constants = Arrays.asList(enumClass.getEnumConstants());

            Arrays.stream(enumClass.getEnumConstants()).map(Enum::name).forEachOrdered(n -> {
                    try {
                            Field field = aClass.getField(n);
                            GraphQLName fieldName = field.getAnnotation(GraphQLName.class);
                            GraphQLDescription fieldDescription = field.getAnnotation(GraphQLDescription.class);
                            Enum constant = constants.stream().filter(c -> c.name().contentEquals(n)).findFirst().get();
                            String name_ = fieldName == null ? n : fieldName.value();
                            builder.value(name_, constant, fieldDescription == null ? name_ : fieldDescription.value());
                        } catch (NoSuchFieldException ignore) {
                        }
                });

            final GraphQLEnumType type = builder.build();
            types.put(typeName, type);
            //noinspection SuspiciousMethodCalls
            processing.remove(type);
            return type;
        }
    }
}
```

**Comment by yarinvak** [comment]
It passes the related tests and I also checked in my project and the enums work (without EnumFunction..).
Can you write what error you get and what code you are trying to run?

**Comment by trunkate** [comment]
hi, here you are with an example:

for the mutation:

```
package test;

import graphql.annotations.annotationTypes.GraphQLDescription;
import graphql.annotations.annotationTypes.GraphQLField;
import graphql.annotations.annotationTypes.GraphQLNonNull;

public class GraphQLMutationWithEnum {

    public enum MyEnum { GOOD, BAD }

    @GraphQLDescription("using a Java enum as input parameter")
    @GraphQLField
    public static MyEnum reverseEnum(@GraphQLNonNull GraphQLMutationWithEnum.MyEnum myEnum) {
        return myEnum.equals(MyEnum.GOOD) ? MyEnum.BAD : MyEnum.GOOD;
    }
}
```

for the query:

```
package test;

import graphql.annotations.annotationTypes.GraphQLField;

public class GraphQLQueryWithEnum {

    @GraphQLField
    public static GraphQLMutationWithEnum.MyEnum good() {
        return GraphQLMutationWithEnum.MyEnum.GOOD;
    }
}

```

the stack trace:

[Java stack trace.txt](https://github.com/graphql-java/graphql-java-annotations/files/1505054/Java.stack.trace.txt)

with the results in GraphiQL:

![mutation](https://user-images.githubusercontent.com/28301357/33256100-e05dad02-d350-11e7-92cb-761881e27d7f.PNG)
![query](https://user-images.githubusercontent.com/28301357/33256101-e07b66b2-d350-11e7-9091-28025418fb62.PNG)

**Comment by trunkate** [comment]
thanks, your latest changes fix it, don't know how.

**Comment by trunkate** [comment]
sorry, I was too optimistic, the mentioned issue is still there

**Comment by guy120494** [comment]
Fixed with #135
